tcpclient/connectionsManagement_test.go

```diff
@@ -13,41 +12,114 @@ import (
+
+	// We use a different subroutine to be able to reach the closeRequest command
+	t.Log("Initiating TCP Connect")
+	go TCPConnect(1, host, port, &wg, ioutil.Discard, statusChannel, closeRequest)
+	if (<-statusChannel).GetConnectionStatus() == ConnectionDialing {
+		t.Log("Connection Dialing")
+	} else {
+		t.Error("Connection failed to dial")
+	}
+	var connectionEstablished = <-statusChannel
+	if connectionEstablished.GetConnectionStatus() == ConnectionEstablished {
+		t.Log("Connection Established")
+	} else {
+		t.Error("Connection failed to establish")
+	}
+	if connectionEstablished.GetTCPProcessingDuration(ConnectionEstablished) != 0 {
+		t.Log("Connection Estalished in ", connectionEstablished.GetTCPProcessingDuration(ConnectionEstablished))
+	} else {
+		t.Error("Connection TCP Processing Duration not consistent")
+	}
+
+	// We ask to close the TCP connection
+	closeRequest <- true
+
+	// Validates wg has been decreased to 0, and next one is making it negative
+	wg.Done()
+	t.Fatal("Should panic")
+}
+
+func TestTCPConnectErrored(t *testing.T) {
+	var host = "127.0.0.1"
+	var port = 55556
+
+	defer func() {
+		err := recover()
+		if err != "sync: negative WaitGroup counter" {
+			t.Fatalf("Unexpected panic: %#v", err)
+		}
+	}()
 	var wg sync.WaitGroup
-	wg.Add(numberConnections)
+	wg.Add(1)
+
+	var statusChannel = make(chan Connection, 2)
+	var closeRequest = make(chan bool)
 
-	for runner := 1; runner <= numberConnections; runner++ {
-		t.Log("Initiating runner # ", strconv.Itoa(runner))
-		go TCPConnect(runner, host, port, &wg, ioutil.Discard, make(chan Connection, numberConnections), make(chan bool))
-		t.Logf("Runner %s initated. Remaining: %s", strconv.Itoa(runner), strconv.Itoa(numberConnections-runner))
+	t.Log("Initiating TCP Connect")
+	TCPConnect(1, host, port, &wg, ioutil.Discard, statusChannel, closeRequest)
+	if (<-statusChannel).GetConnectionStatus() == ConnectionDialing {
+		t.Log("Connection Dialing")
+	} else {
+		t.Error("Connection failed to dial")
 	}
 
-	t.Log("Waiting runners to finish")
-	time.Sleep(time.Second)
+	var connectionErrored = <-statusChannel
 
-	// Marking wait groups as done after a second does not make sense...
-	// and in case one client finishes before this loop being executed (because of an error?)
-	// we will get an exception (panic: sync: negative WaitGroup counter [recovered])
+	if connectionErrored.GetConnectionStatus() == ConnectionError {
+		t.Log("Connection Errored")
+	} else {
+		t.Error("Connection not errored")
+	}
+	if connectionErrored.GetTCPProcessingDuration(ConnectionError) != 0 {
+		t.Log("Connection Errored in ", connectionErrored.GetTCPProcessingDuration(ConnectionError))
+	} else {
+		t.Error("Connection TCP Processing Duration not consistent")
+	}
 
-	//for runner := 1; runner <= numberConnections; runner++ {
-	//	t.Log("Closing runner #", strconv.Itoa(runner))
-	//	wg.Done()
-	//}
+	// Validates wg has been decreased to 0, and next one is making it negative
+	wg.Done()
```

Comment:
I don't understand how this works... TCPConnect should have decreased the wg, so this means an effective error, right? But running the test suit does not complain...

Reply:
Yup, you are right, this is an error which is captured by https://github.com/dachad/tcpgoon/blob/testing-tcpclient/tcpclient/connectionsManagement_test.go#L31.
So, basically, this `wg.Done()` means that we are forcing an error `sync: negative WaitGroup counter` and that's what's expected to pass the test